### PR TITLE
Add parameter period to set the output frequency of XSens device

### LINKED ISSUE #33: Add the posibility to stream IMU data at the desired rate
The current implementation of the [`xsensmt` device](https://github.com/robotology/yarp-device-xsensmt/tree/master/xsensmt) allows for streaming at ~100Hz.

Quoting the comment from @traversaro in another location:

> ## IMU
> 
> > For some reason, the only file called is https://github.com/ami-iit/robots-configuration/blob/devel_iCubGenova04_higherFreq/iRonCub-Mk1/hardware/inertials/waist-inertial.xml which also open the waist IMU port
> 
> Some YARP Network Wrapper Server permit to open hardware devices via the `subdevice` parameter, and this is what is done in this case, see https://github.com/ami-iit/robots-configuration/blob/e1d22102852e72745a748635a9f0898be57decd9/iRonCub-Mk1/hardware/inertials/waist-inertial.xml#L8  . However, the period parameter that is read by the `inertial` YARP device is not read by the `xsensmt`  YARP device. The paremeters supported are not documented it seems, but can be found in https://github.com/robotology/yarp-device-xsensmt/blob/6de80ae7d3a67eb315d4972c2e2023d73cf2dc5a/xsensmt/XsensMT.cpp#L91 , and no `period`  parameter is read. The data is read by the `sensorReadLoop`, that is continuously receiving data from the serial connection to the sensor:
> * https://github.com/robotology/yarp-device-xsensmt/blob/6de80ae7d3a67eb315d4972c2e2023d73cf2dc5a/xsensmt/XsensMT.cpp#L217
> * https://github.com/robotology/yarp-device-xsensmt/blob/6de80ae7d3a67eb315d4972c2e2023d73cf2dc5a/xsensmt/XsensMT.cpp#L250
> * https://github.com/robotology/yarp-device-xsensmt/blob/6de80ae7d3a67eb315d4972c2e2023d73cf2dc5a/xsensmt/XsensMT.cpp#L324
> 
> If the data is coming at ~100 Hz, probably it is because the Xsens sensor is sending it at that rate: probably we should modify the driver to permit it to ask to the Xsens sensor to send it at higher frequency, probably modifying around this part of the code: https://github.com/robotology/yarp-device-xsensmt/blob/6de80ae7d3a67eb315d4972c2e2023d73cf2dc5a/xsensmt/XsensMT.cpp#L172. Se also Table 2 in http://web.archive.org/web/20211212143408/https://www.xsens.com/hubfs/Downloads/Manuals/MT_Low-Level_Documentation.pdf, it seems that with the profile in which we set the sensor the expected output is indeed 100 Hz. We could open an issue on this at https://github.com/robotology/yarp-device-xsensmt/issues .

It would be nice to have the possibility to make the period a configurable parameter.

### PR DESCRIPTION
**Pull request by HosameldinMohamed**
Fixes #33.

### CODE REVIEW

File: xsensmt/XsensMT.cpp [hunk 1]
```diff
@@ -108,13 +108,29 @@ bool XsensMT::open(yarp::os::Searchable &config)
...
     }
 
+    yDebug() << "xsensmt - test - xsensmt_period, check xsensmt_period: " << config.check("xsensmt_period");
+    yDebug() << "xsensmt - test - xsensmt_period, xsensmt_period is string: " << config.find("xsensmt_period").isString();
+    yDebug() << "xsensmt - test - xsensmt_period, xsensmt_period is int: " << config.find("xsensmt_period").isInt32();
+    yDebug() << "xsensmt - test - xsensmt_period, xsensmt_period is float: " << config.find("xsensmt_period").isFloat64();
+
+    if (config.check("xsensmt_period") && ( config.find("xsensmt_period").isInt32() || config.find("xsensmt_period").isFloat64()))
+    {
+        m_outputPeriod = config.find("xsensmt_period").asFloat64();
```
Comment (traversaro) [inline xsensmt/XsensMT.cpp:118]: Are you sure that this works as intended if xsensmt_period is a integer?

File: xsensmt/XsensMT.cpp [hunk 2]
```diff
@@ -108,13 +108,29 @@ bool XsensMT::open(yarp::os::Searchable &config)
...
+    }
+    else
+    {
+        m_outputPeriod = 10; // 10ms
```
Comment (traversaro) [inline xsensmt/XsensMT.cpp:122]: The latest style for most YARP devices is to use seconds to represent periods, see for example https://www.yarp.it//v3.5/classControlBoard__nws__yarp.html#details . Even if some icub devices still use milliseconds, I would try to be consistent with the new devices and use seconds here.

File: xsensmt/XsensMT.cpp [hunk 3]
```diff
@@ -108,13 +108,29 @@ bool XsensMT::open(yarp::os::Searchable &config)
         yWarning() << "xsensmt -  Parameter \"frame_name\" not set. Using the same value as \"sensor_name\" for this parameter.";
     }
 
+    yDebug() << "xsensmt - test - xsensmt_period, check xsensmt_period: " << config.check("xsensmt_period");
+    yDebug() << "xsensmt - test - xsensmt_period, xsensmt_period is string: " << config.find("xsensmt_period").isString();
+    yDebug() << "xsensmt - test - xsensmt_period, xsensmt_period is int: " << config.find("xsensmt_period").isInt32();
+    yDebug() << "xsensmt - test - xsensmt_period, xsensmt_period is float: " << config.find("xsensmt_period").isFloat64();
+
+    if (config.check("xsensmt_period") && ( config.find("xsensmt_period").isInt32() || config.find("xsensmt_period").isFloat64()))
```
Comment (traversaro) [inline xsensmt/XsensMT.cpp:116]: Once you are ready, remember to add the parameter in https://github.com/robotology/yarp-device-xsensmt/blob/b31cba62d2b3f31a3f60d8353e658005b77fe7d4/xsensmt/XsensMT.h#L41 . Probably it could make sense to move that documentation in the README.md to increase its visibility, but that is not related to this PR.

File: xsensmt/XsensMT.cpp (hunk 1 above)
Comment (HosameldinMohamed) [reply xsensmt/XsensMT.cpp:118]: I ran some checks, if I set `xsensmt_period` to an integer value (e.g. `2`), `config.find("xsensmt_period").isInt32()` returns `true` and `config.find("xsensmt_period").isFloat()` returns `false`. The opposite happens when I set a floating number (e.g. `2.0`)

`config.find("xsensmt_period").isString()` returned `false` in both cases.

That's why I put an `OR`.

On the other hand, `m_outputPeriod = config.find("xsensmt_period").asFloat64();` works as expected. If I set `2` for example, it's stored as `2.00`, and the float value is returned correctly too (with 2 decimal points).

Comment (traversaro) [reply xsensmt/XsensMT.cpp:118]: Perfect!

**Comment by HosameldinMohamed** [comment]
## Behavior test

I am testing with the following command (selecting the period)

```sh
yarpdev --device inertial --name /icub/xsens_inertial --period 0.002 --subdevice xsensmt --serial /dev/ttyXsens --xsensmt_period 2.7
```

It seems that the SDK is able to configure the IMU with the period I am selecting.

When I select `xsensmt_period=2` (`500Hz`) I got a continuous stream of the warning

```
[WARNING] xsensmt: Missing magnetometer message, skipping magneto update.
```

This agrees to what @gabrielenava and @affafjunaidMOMIN found in the manual that the max output frequency is `400Hz` (`2.5ms`) for `MTi-300`.

But I get the same when I select the period `2.5ms`.

Instead when selecting the period `2.6ms`, I get the same warning for all the quantities (orientation, ACC, gyro, magneto), but **only once in a while**.

**Comment by traversaro** [comment]
You can also think of having separate parameters for each quantity.

**Comment by HosameldinMohamed** [comment]
> This agrees with what @gabrielenava and @affafjunaidMOMIN found in the manual that the max output frequency is `400Hz` (`2.5ms`) for `MTi-300`.

**Update**: the **actual** max period we can get is around `5ms` (`200Hz`). This is checked by running the following:

- code at the current status: https://github.com/robotology/yarp-device-xsensmt/pull/34/commits/5fa089934b10bff33f2d6c531f702a7868ddb3be

- Running 

```
yarpdev --device inertial --name /icub/xsens_inertial --period 0.003 --subdevice xsensmt --serial /dev/ttyXsens --xsensmt_period 3
```
Then `yarpdatadumper`

```
yarpdatadumper --name /dumper/xsens_inertial --connect /icub/xsens_inertial --rxTime --txTime
```

And checking the sender time of the logged data.

If I understood correctly from the code, the port publishes new line in the YARP port as soon as it gets a new data samples from the IMU. So the port sending time is indicative of the frequency of IMU (@traversaro can you confirm?)

So it seems that the IMU output frequency saturates at `200Hz`. In the manual it says `400Hz`, except for the Magnetometer (`100Hz`). See table `16` in page `32` of [the manual](http://web.archive.org/web/20211212143408/https://www.xsens.com/hubfs/Downloads/Manuals/MT_Low-Level_Documentation.pdf).

**Comment by traversaro** [comment]
> If I understood correctly from the code, the device publishes new line in the YARP port as soon as it gets a new data samples from the IMU. So the port sending time is indicative of the frequency of IMU (@traversaro can you confirm?)

No, the hardware device make the new measurements available in the C++ methods such as `getOrientationSensorMeasureAsRollPitchYaw` or `read`  as soon as they arrive. Then the Network Wrapper Server device that is attached to the `xsensmt` device, such as `inertial`, calls `read` and publish it on the YARP port with the frequency given by its own thread.

**Comment by HosameldinMohamed** [comment]
> No, the hardware device make the new measurements available in the C++ methods such as `getOrientationSensorMeasureAsRollPitchYaw` or `read` as soon as they arrive. Then the Network Wrapper Server device that is attached to the `xsensmt` device, such as `inertial`, calls `read` and publish it on the YARP port with the frequency given by its own thread.

Ah okay, then it is strange that port is published in not more than `5ms` even if I put a lesser value. Maybe there's a saturation implemented in the Network Wrapper Service?

**Comment by HosameldinMohamed** [comment]
> Ah okay, then it is strange that port is published in not more than `5ms` even if I put a lesser value. Maybe there's a saturation implemented in the Network Wrapper Service?

I did a quick check, I am suspecting that https://github.com/robotology/yarp/blob/5804efb739319aee0bf1bfe408d80b31e7195238/src/devices/ServerInertial/ServerInertial.cpp#L284 is responsible.

**Comment by traversaro** [comment]
> > No, the hardware device make the new measurements available in the C++ methods such as `getOrientationSensorMeasureAsRollPitchYaw` or `read` as soon as they arrive. Then the Network Wrapper Server device that is attached to the `xsensmt` device, such as `inertial`, calls `read` and publish it on the YARP port with the frequency given by its own thread.
> 
> Ah okay, then it is strange that port is published in not more than `5ms` even if I put a lesser value. Maybe there's a saturation implemented in the Network Wrapper Service?

Good point, we can check in the code: <insert link>.

**Comment by HosameldinMohamed** [comment]
> Good point, we can check in the code: `<insert link>`.

I didn't understand,, wrong place?

**Comment by traversaro** [comment]
> > Good point, we can check in the code: `<insert link>`.
> 
> I didn't understand,, wrong place?

Sorry, I wanted to insert the link but I am lagging behind.

**Comment by HosameldinMohamed** [comment]
this? https://github.com/robotology/yarp-device-xsensmt/pull/34#issuecomment-1138599534

**Comment by traversaro** [comment]
> this? [#34 (comment)](https://github.com/robotology/yarp-device-xsensmt/pull/34#issuecomment-1138599534)

Yes!

**Comment by traversaro** [comment]
@HosameldinMohamed please write me on Teams if you want to quickly align on this to proceed with this PR.

**Comment by HosameldinMohamed** [comment]
> Ah okay, then it is strange that port is published in not more than `5ms` even if I put a lesser value. Maybe there's a saturation implemented in the Network Wrapper Service?

I am testing with the `fakeIMU` device wrapped by the `inertial` device, using the command:

```sh
yarpdev --device inertial --name /fakeHa --period 0.001 --subdevice fakeIMU
```

The response is as expected, and the actual period is close to `1ms`. I checked the `timestamps` using `yarpdatadumper`.

File: xsensmt/XsensMT.cpp (hunk 3 above)
Comment (HosameldinMohamed) [reply xsensmt/XsensMT.cpp:111]: Ok, add the parameter to the table in https://github.com/robotology/yarp-device-xsensmt/pull/34/commits/28bdd1b1c310f84e74e5d67c6032cc3c6f8d07a6

File: xsensmt/XsensMT.cpp (hunk 2 above)
Comment (HosameldinMohamed) [reply xsensmt/XsensMT.cpp]: Ok now the expected period is in seconds https://github.com/robotology/yarp-device-xsensmt/pull/34/commits/28bdd1b1c310f84e74e5d67c6032cc3c6f8d07a6
